core/distributed/preconditioner/schwarz.cpp

```diff
@@ -130,15 +182,52 @@ void Schwarz<ValueType, LocalIndexType, GlobalIndexType>::generate(
             "Requires either a generated solver or an solver factory");
     }
 
-    if (parameters_.local_solver) {
-        this->set_solver(gko::share(parameters_.local_solver->generate(
-            as<experimental::distributed::Matrix<
-                ValueType, LocalIndexType, GlobalIndexType>>(system_matrix)
-                ->get_local_matrix())));
+    if ((parameters_.coarse_level && !parameters_.coarse_solver) ||
+        (!parameters_.coarse_level && parameters_.coarse_solver)) {
+        GKO_INVALID_STATE(
+            "Requires both coarse solver and coarse level to be set.");
+    }
+    auto dist_mat =
+        as<experimental::distributed::Matrix<ValueType, LocalIndexType,
+                                             GlobalIndexType>>(system_matrix);
+    if (parameters_.coarse_weight > 0 && parameters_.coarse_weight <= 1) {
+        this->local_weight_ = gko::initialize<matrix::Dense<ValueType>>(
+            {ValueType{1} - ValueType{parameters_.coarse_weight}},
+            this->get_executor());
+        this->coarse_weight_ = gko::initialize<matrix::Dense<ValueType>>(
+            {ValueType{parameters_.coarse_weight}}, this->get_executor());
+    } else {
+        this->local_weight_ = gko::initialize<matrix::Dense<ValueType>>(
+            {ValueType{1.0}}, this->get_executor());
+        this->coarse_weight_ = gko::initialize<matrix::Dense<ValueType>>(
+            {ValueType{1.0}}, this->get_executor());
+    }
 
+    if (parameters_.local_solver) {
+        this->set_solver(gko::share(
+            parameters_.local_solver->generate(dist_mat->get_local_matrix())));
     } else {
         this->set_solver(parameters_.generated_local_solver);
     }
+
+    if (parameters_.coarse_level && parameters_.coarse_solver) {
+        this->coarse_level_ =
+            share(parameters_.coarse_level->generate(system_matrix));
+        if (as<multigrid::MultigridLevel>(this->coarse_level_)) {
+            auto coarse = as<multigrid::MultigridLevel>(this->coarse_level_)
+                              ->get_coarse_op();
+            // Cannot guarantee that coarse op is not nullptr
```

Comment:
should it throw the error rather than make the operation silent?

Reply:
I guess that depends on whether we consider generation of a coarse level valid with empty triplet (R, A, P = nullptr)

Reply:
I think we should throw an error now. The previous issue here is caused by that the test does not initialize the coarse solver and coarse level properly. The slight operation on this will give user wrong performance thought. If someone comes to use for this behavior later, we can add another parameter for that but not the default behavior.

Reply:
No, the issue is that when pgm (or MultigridLevel) is generated with an empty matrix, then the coarse op is null. We should probably ensure that the triplet generated by MultigridLevel cannot be null. If we throw here then it means that the Schwarz preconditioner will not be in a valid state.